### PR TITLE
clm lake fix: pressure is not density + #142 and #148

### LINKED ISSUE #140: CLM Lake uses near-surface pressure in place of near-surface density
# Description

The `rho0` variable in CLM Lake is supposed to be the near-surface density. Instead, the model uses near-surface pressure. One RRFS case crashed due to this.

Bug identified and fixed by @tanyasmirnova 

## Steps to Reproduce

Run this test case with your favorite version of ufs-weather-model.

HERA: /scratch1/BMC/zrtrr/Samuel.Trahan/lakes/lake-crash-2023120400

## Additional Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* Machine - Hera
* Compiler - Intel
* Suite Definition File or Scheme - FV3_HRRR_gf
* Reference other issues or PRs in other repositories that this is related to, and how they are related.

### LINKED ISSUE #141: crash when running GFS PBL/sfclay and clm lake in RRFS ensemble
# Description

A couple of RRFS ensemble members crashed with GFS PBL (TKE-EDMF)/sfclay and clm lake combined. This happened when lake water was frozen to lake ice in clm lake.
  
## Steps to Reproduce

1. perform a forecast with GFS PBL/sfclayer combined with clm lake
2. wait for lake water turning into lake ice, which could happen in winter time further north
3. model crashed with AVOST error message from RUC LSM

## Additional Context

The reason for the crash is because clm lake is called between surface layer scheme and PBL scheme. If there is a state change in clm lake, particularly water turning into ice, stability related variables over ice will be missing and cause the crash.
Please see the attached slides for more information.
   
[clm_lake_bug_fix.pptx](https://github.com/ufs-community/ccpp-physics/files/13705065/clm_lake_bug_fix.pptx)

A PR will be submitted soon.
## Output

Please include any relevant log files, screenshots or other output here.

### LINKED ISSUE #143: RUC LSM snow depth accumulation diagnostic should only include snow fall with variable density
## Description
Quoting @tanyasmirnova 

> "[We need a] change in the diagnostic of the snow depth accumulation. The old diagnostic included snow melting, but in the new version it includes only the snow fall with variable density.
> "This was decided after consultations with people from the Weather Service."

## Solution
Modify the RUC LSM to calculate the diagnostic differently.

### LINKED ISSUE #144: RUC LSM snow melting and sublimation should not be multiplied by snow fraction
# Description

Quoting @tanyasmirnova 

> "The old [snowfall accumulation] diagnostics that included snow melting in the computation of snowfallacc had a bug in the computation of snow melt. Snow melt and snow sublimation should not be multiplied by snowfrac. It is also only the diagnostic variables."

## Steps to Reproduce

You'll see lines like this in module_sf_ruclsm.f90:

- `smelt = smelt * snowfrac`

Which scale diagnostic by snow fraction.

## Additional Context

Problem was discovered during RRFS testing

### LINKED ISSUE #145: turn off separate treatment of snow covered and snow-free portions of the grid cell in RUC LSM when fresh snow falls on the snow-free ground
## Description
RRFS testing shows better model performance if one makes a change to RUC LSM. Specifically, turning off separate treatment of snow-covered and snow-free portions of a grid cell when snow falls on the snow-free surface. This change matches the code used in the operational HRRR.

## Solution
`snow_mosaic=0` in one line of module_sf_ruclsm.f90 for the case of fresh snow falling on snow-free surface. Otherwise, still use snow_mosaic = 1.

### LINKED ISSUE #146: If input_lakedepth is lost between restarts, CLM Lake model will crash
# Description

If CLM Lake is used in FV3 as intended, `input_lakedepth` should never be 0. The world is not always ideal. Sometimes there may be a change in orography between cycles of a cycled model. Or, something may go wrong in the model restart procedure. This can lead to `input_lakedepth=0` on some points.

When that happens, there is a divide by zero in the clm lake mode.

## Steps to Reproduce

1. Fill `input_lakedepth` with 0 in a restart file.
2. Restart the model.
3. Witness the divide by zero error.

## Additional Context

This was discovered during RRFS parallels.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
***PRs #142 and #148 are combined into this one***

The CLM Lake Model was using near-surface pressure in place of near-surface density. This corrects the code to user near-surface density instead.

- fixes #140
- fixes #146

These two pull requests are included in this one. They can be closed once this one is merged.
- closes #142 
- closes #148 

Due to including #148:
- fixes #141 

And from #142:
- fixes #143 
- fixes #144 
- fixes #145

### CODE REVIEW

**Comment by tanyasmirnova** [comment]
@SamuelTrahanNOAA This PR also fixes the crash with the use of the CLM lake model caused by a lake point with the lake depth < 10 cm.

**Review by tanyasmirnova — APPROVED** [review]

File: physics/clm_lake.f90 [hunk 1]
```diff
@@ -238,7 +238,7 @@ subroutine calculate_z_dz_lake(i,input_lakedepth,clm_lakedepth,z_lake,dz_lake)
       real(kind_lake) :: dz_lake(nlevlake) ! layer thickness for lake (m)
       real(kind_lake) :: depthratio
 
-      if (input_lakedepth(i) == spval) then
+      if (input_lakedepth(i) == spval .or. input_lakedepth(i) < 0.1) then
```
Comment (dustinswales) [inline physics/clm_lake.f90:241]: Can you add a comment for this new parameter? (minimum lake depth in meters?)

**Review by dustinswales — APPROVED** [review]
@SamuelTrahanNOAA Looks good. Just a small request to comment on a new parameter

File: physics/clm_lake.f90 (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply physics/clm_lake.f90:241]: This is a workaround for a disturbing bug we have been unable to track down in the RRFS parallel. The lakedepth doesn't match the lake mask. I suspect something went wrong in the cycling, or the orography data was changed during the parallel.

Comment (SamuelTrahanNOAA) [reply physics/clm_lake.f90:241]:
```f90
      if (input_lakedepth(i) == spval .or. input_lakedepth(i) < 0.1) then
        ! This is a safeguard against:                                                                                                                                                                                                       
        ! 1. missing in the lakedepth database (== spval)                                                                                                                                                                                    
        ! 2. errors in model cycling or unexpected changes in the orography database (< 0.1)                                                                                                                                                 
```

Comment (dustinswales) [reply physics/clm_lake.f90:241]: Thanks Sam!

**Review by grantfirl — APPROVED** [review]

**Comment by SamuelTrahanNOAA** [comment]
This PR fixes another bug I forgot to mention. I've added an issue for it:

- #146

**Review by Qingfu-Liu — APPROVED** [review]

**Review by tanyasmirnova — APPROVED** [review]

**Comment by SamuelTrahanNOAA** [comment]
@JiliDong-NOAA - Please confirm that I merged your changes into this branch correctly.

**Comment by JiliDong-NOAA** [comment]
> @JiliDong-NOAA - Please confirm that I merged your changes into this branch correctly.

merge looks correct.

**Comment by jkbk2004** [comment]
All tests were finished at https://github.com/ufs-community/ufs-weather-model/pull/2044. @grantfirl @dustinswales @tanyasmirnova @Qingfu-Liu Can you merge this pr?

**Comment by dustinswales** [comment]
Done. Note that the [CI tests will fail](https://github.com/ufs-community/ccpp-physics/actions/runs/7290178396/job/19866373420) until [fv3atm #736](https://github.com/NOAA-EMC/fv3atm/pull/736) is merged

**Pull request merged**